modules/schema/src/main/java/org/apache/ignite/internal/schema/BinaryTupleComparatorUtils.java

```diff
@@ -0,0 +1,305 @@
+
+            case BYTES:
+                return Arrays.compareUnsigned(tuple1.bytesValue(index1), tuple2.bytesValue(index2));
+
+            case UUID:
+                return tuple1.uuidValue(index1).compareTo(tuple2.uuidValue(index2));
+
+            case STRING:
+                return tuple1.stringValue(index1).compareTo(tuple2.stringValue(index2));
+
+            case DECIMAL:
+                BigDecimal numeric1 = tuple1.decimalValue(index1, Integer.MIN_VALUE);
+                BigDecimal numeric2 = tuple2.decimalValue(index2, Integer.MIN_VALUE);
+
+                return numeric1.compareTo(numeric2);
+
+            case TIMESTAMP:
+                return tuple1.timestampValue(index1).compareTo(tuple2.timestampValue(index2));
+
+            case DATE:
+                return tuple1.dateValue(index1).compareTo(tuple2.dateValue(index2));
+
+            case TIME:
+                return tuple1.timeValue(index1).compareTo(tuple2.timeValue(index2));
+
+            case DATETIME:
+                return tuple1.dateTimeValue(index1).compareTo(tuple2.dateTimeValue(index2));
+
+            default:
+                throw new IllegalArgumentException(format("Unsupported column type in binary tuple comparator. [type={}]", typeSpec));
+        }
+    }
+
+    static boolean isFlagSet(ByteBuffer tuple, int flag) {
+        return (tuple.get(0) & flag) != 0;
+    }
+
+    static int equalityFlag(ByteBuffer tuple) {
+        return isFlagSet(tuple, EQUALITY_FLAG) ? 1 : -1;
+    }
+
+    /**
+     * Compares a value in a binary tuple, interpreted as a string, with a given string.
+     * The comparison can be performed as case-sensitive or case-insensitive.
+     * The method first attempts a fast comparison for ASCII sequences and falls back
+     * to Unicode comparison if non-ASCII characters are detected.
+     *
+     * @param tuple The BinaryTupleReader containing the tuple to be compared.
+     * @param colIndex The column index in the tuple to retrieve the value for comparison.
+     * @param cmp The string to compare the value in the tuple against.
+     * @param ignoreCase Flag indicating whether the comparison should ignore case differences.
+     * @return 0 if the strings are equal, a negative value if the tuple string is lexicographically
+     *         less than the given string, or a positive value if it is greater.
+     */
+    static int compareAsString(BinaryTupleReader tuple, int colIndex, String cmp, boolean ignoreCase) {
+        tuple.seek(colIndex);
+        int begin = tuple.begin();
+        int end = tuple.end();
+
+        ByteBuffer buf = tuple.byteBuffer();
+        int fullSrtLength = end - begin;
+        int trimmedSize = Math.min(fullSrtLength, buf.capacity() - begin);
+        byte[] bytes = new byte[trimmedSize];
+
+        buf.duplicate().position(begin).limit(begin + trimmedSize).get(bytes);
+
+        char[] cmpArray = cmp.toCharArray();
```

Comment:
I'm not really convinced that we need to copy these arrays, especially in case of `cmpArray`. I presume that version without copying was slower, is that the case? Please add some small comment

Reply:
Yes, access to any byte by index for an indirect array is longer than access by index in a pure array. I benchmarked it with and without copying; approaching with a full copy in heap is preferable.

Reply:
Added comment.